### ISSUE TITLE
Broken shutdown flow in sync

### ISSUE THREAD

**Issue by lrettig** [spacemeshos/go-spacemesh#2178]
## Description

`TestAppTestSuite/TestMultipleNodes` is currently failing with a panic that appears to be happening during the shutdown flow at the end of the test:

```
panic: send on closed channel

goroutine 10872 [running]:
github.com/spacemeshos/go-spacemesh/sync.(*fetchQueue).addToPending(0xc000be32b0, 0xc00fcf9680, 0x6, 0x6, 0xc00fcf9680, 0x6, 0x6)
	/home/runner/work/go-spacemesh/go-spacemesh/sync/fetch_queue.go:113 +0x425
github.com/spacemeshos/go-spacemesh/sync.(*fetchQueue).addToPendingGetCh(0xc000be32b0, 0xc00fcf9680, 0x6, 0x6, 0xc013850f60)
	/home/runner/work/go-spacemesh/go-spacemesh/sync/fetch_queue.go:95 +0x49
github.com/spacemeshos/go-spacemesh/sync.(*fetchQueue).handle(0xc000be32b0, 0xc00fcf9440, 0x6, 0x6, 0x1, 0x1, 0x49c8e5)
	/home/runner/work/go-spacemesh/go-spacemesh/sync/fetch_queue.go:140 +0xa5
github.com/spacemeshos/go-spacemesh/sync.atxQueue.HandleAtxs(0xc016e4fce0, 0xc000116ee0, 0xc0144666f0, 0x1272b00, 0xc01965a390, 0xc01965a4e0, 0xc012386cc0, 0xc0142d2b30, 0xc016e4fd40, 0x1094186, ...)
	/home/runner/work/go-spacemesh/go-spacemesh/sync/fetch_queue.go:252 +0x1bf
github.com/spacemeshos/go-spacemesh/sync.(*Syncer).fetchAllReferencedAtxs(0xc011ca6a00, 0xc01dac7600, 0xc013a8b420, 0x20)
	/home/runner/work/go-spacemesh/go-spacemesh/sync/syncer.go:699 +0x1a9
github.com/spacemeshos/go-spacemesh/sync.(*Syncer).fetchBlockDataForValidation(0xc011ca6a00, 0xc01dac7600, 0x1245a40, 0xc0000491c0)
	/home/runner/work/go-spacemesh/go-spacemesh/sync/syncer.go:710 +0x43
github.com/spacemeshos/go-spacemesh/sync.(*Syncer).blockSyntacticValidation(0xc011ca6a00, 0xc01dac7600, 0xdbc3abe5b2e58ffe, 0xc054ea2df7, 0x0, 0x1245a40, 0xc0000491c0, 0x0, 0x0, 0x0)
	/home/runner/work/go-spacemesh/go-spacemesh/sync/syncer.go:715 +0x50
github.com/spacemeshos/go-spacemesh/sync.(*BlockListener).handleBlock(0xc01b688e40, 0x1265540, 0xc01a48d500)
	/home/runner/work/go-spacemesh/go-spacemesh/sync/block_listener.go:123 +0xc91
github.com/spacemeshos/go-spacemesh/sync.(*BlockListener).listenToGossipBlocks.func1(0xc01b688e40, 0x1265540, 0xc01a48d500)
	/home/runner/work/go-spacemesh/go-spacemesh/sync/block_listener.go:75 +0xa1
created by github.com/spacemeshos/go-spacemesh/sync.(*BlockListener).listenToGossipBlocks
	/home/runner/work/go-spacemesh/go-spacemesh/sync/block_listener.go:68 +0x151
make: *** [test-only-app-test] Error 1
FAIL	github.com/spacemeshos/go-spacemesh/cmd/node	428.186s
FAIL
Makefile:148: recipe for target 'test-only-app-test' failed
```

See https://github.com/spacemeshos/go-spacemesh/runs/1324034214?check_suite_focus=true

CC @antonlerner 

## Steps To Reproduce
Run this test. Note that it only appears to be failing in CI; @noamnelke reported that it's passing when he runs it locally.

## Environment
Please complete the following information:
 - OS: GitHub Actions CI ubuntu-latest
 - Node Version: 21acf374e194e1a5fbfc5de311ddf39ade8b78d0

**Comment by lrettig** [comment]
This failed again, after including the fix from #2182 :(

https://github.com/spacemeshos/go-spacemesh/runs/1334509150?check_suite_focus=true

**Comment by lrettig** [comment]
Here's my diagnosis of what's going on. The way we handle gossip channels is a little messy. Ideally, these channels would be closed as part of the p2p shutdown process, but right now they're not, and this is marked as a to do:

https://github.com/spacemeshos/go-spacemesh/blob/85dc316082f0c375944b7f88387ee1a943354dcf/p2p/switch.go#L431-L438

It would be cleaner if these channels were closed at the source and we could be sure that there were no further incoming gossip messages, but that's not the case and instead we try to shut things down from the listener side. When the BlockListener.Close() is called, it closes an exit channel that's a signal to stop listening to gossip, and it makes sure there aren't any blocks currently being processed before exiting. That's good. The problem is that there may already be an incoming gossip message queued up on the gossip channel, and when this select{} block next loops, it's nondeterministic whether it will see the closed exit channel or the incoming message first - hence this test failing unpredictably:

https://github.com/spacemeshos/go-spacemesh/blob/85dc316082f0c375944b7f88387ee1a943354dcf/sync/block_listener.go#L56-L78

I added a workaround in #2186, but we should rethink how we handle closing gossip channels.
